File: galleries/08-file-system/src/main.py
"""브라우저 안의 파일시스템을 파이썬 쪽에서 다룬다."""

import io
import zipfile
from pathlib import Path

WORK = Path("/work")
PERSIST = Path("/persist")


def setup() -> None:
    """작업 폴더를 만든다. /persist 는 자바스크립트가 IDBFS 로 마운트해 둔다."""
    WORK.mkdir(exist_ok=True)


def list_tree(root: str) -> list[tuple[str, int, bool]]:
    """폴더 하나를 훑어 (경로, 크기, 폴더인가) 목록을 만든다."""
    base = Path(root)
    if not base.exists():
        return []

    found: list[tuple[str, int, bool]] = []
    for path in sorted(base.rglob("*")):
        is_dir = path.is_dir()
        size = 0 if is_dir else path.stat().st_size
        found.append((str(path), size, is_dir))
    return found


def describe(path: str) -> str:
    """올라온 파일을 파이썬이 읽어 무엇인지 말해 준다.

    앞부분만 보여 주면 되니 통째로 읽지 않는다. 몇십 MB 짜리를 올려도 이 함수 때문에
    한 벌이 더 잡히는 일이 없다.
    """
    target = Path(path)
    size = target.stat().st_size
    with target.open("rb") as handle:
        head = handle.read(240)

    # 240바이트에서 자르면 글자 중간이 끊길 수 있다. 그 자리를 물음표로 바꾸고
    # 글자 단위로 다시 센다. errors 를 안 주면 평범한 한글 파일이 16진수로 나온다.
    text = head.decode("utf-8", errors="replace")
    printable = sum(1 for ch in text if ch.isprintable() or ch in "\n\t")
    if printable < len(text) * 0.8:
        return f"{target.name}: {size}바이트\n앞부분(글자가 아닌 것 같아 16진수로): {head[:40].hex(' ')}"
    return f"{target.name}: {size}바이트\n앞부분: {text[:60]}"


def make_report(note: str) -> str:
    """파이썬이 파일을 하나 만든다. 이걸 내려받게 된다."""
    target = WORK / "report.txt"
    lines = [
        "이 파일은 브라우저 안의 파이썬이 만들었습니다.",
        f"남긴 말: {note}",
        "",
        "작업 폴더에 있는 것:",
        *(
            f"  {p.name} ({p.stat().st_size}바이트)"
            for p in sorted(WORK.iterdir())
            if p.is_file()
        ),
    ]
    target.write_text("\n".join(lines), encoding="utf-8")
    return str(target)


def write_persistent(name: str, text: str) -> str:
    """영속 폴더에 쓴다. 실제로 남기려면 자바스크립트가 syncfs 를 불러야 한다."""
    target = PERSIST / name
    target.write_text(text, encoding="utf-8")
    return str(target)


def make_zip() -> bytes:
    """압축 파일을 하나 만들어 바이트로 돌려준다. 이걸 unpackArchive 가 푼다."""
    buffer = io.BytesIO()
    with zipfile.ZipFile(buffer, "w") as archive:
        archive.writestr("풀린것/하나.txt", "첫 번째 파일")
        archive.writestr("풀린것/둘.txt", "두 번째 파일")
        archive.writestr("풀린것/안쪽/셋.txt", "폴더 안의 파일")
    return buffer.getvalue()

File: galleries/08-file-system/src/main.js
// 08. 파일 다루기
//
// 새로 배우는 것은 넷이다.
//   pyodide.FS 로 자바스크립트에서 파일을 넣고 꺼낸다
//   올린 파일을 파이썬에게 주고, 파이썬이 만든 것을 내려받는다
//   MEMFS 는 새로고침하면 사라진다
//   IDBFS 는 남지만 syncfs 를 불러야 한다

import { ensureSupport } from '../../_shared/support.js';
import { getPyodide, renderPythonError, showLoading } from '../../_shared/pyodide.js';
import { formatBytes } from '../../_shared/metrics.js';

const WORK = '/work';
const PERSIST = '/persist';
const UNPACKED = '/unpacked';
const ROOTS = [WORK, PERSIST, UNPACKED];

const bootBox = document.querySelector('#boot');
const refreshButton = document.querySelector('#refresh');
const treeBody = document.querySelector('#tree tbody');
const uploadInput = document.querySelector('#upload');
const uploadResult = document.querySelector('#upload-result');
const noteInput = document.querySelector('#note');
const downloadButton = document.querySelector('#download');
const downloadResult = document.querySelector('#download-result');
const saveSyncButton = document.querySelector('#save-sync');
const saveNoSyncButton = document.querySelector('#save-nosync');
const reloadButton = document.querySelector('#reload');
const persistResult = document.querySelector('#persist-result');
const unpackButton = document.querySelector('#unpack');
const unpackResult = document.querySelector('#unpack-result');

let pyGlobals = null;

if (ensureSupport()) {
  start().catch((error) => renderPythonError(bootBox, error));
}

async function start() {
  bootBox.replaceChildren();
  const done = showLoading(bootBox);

  let pyodide;
  try {
    pyodide = await getPyodide();
  } catch (error) {
    done();
    renderPythonError(bootBox, error);
    return;
  }

  // 저장소가 막힌 브라우저가 있다. 그때도 나머지 기능은 쓸 수 있어야 하므로 따로 감싼다.
  let persistReady = true;
  try {
    // IDBFS 를 붙이는 것은 자바스크립트 쪽 일이다. 파이썬에서는 그냥 폴더로 보인다.
    pyodide.FS.mkdirTree(PERSIST);
    pyodide.FS.mount(pyodide.FS.filesystems.IDBFS, {}, PERSIST);
    // 켤 때 한 번 읽어 와야 지난번에 저장한 것이 보인다.
    await syncfs(pyodide, true);
  } catch (error) {
    persistReady = false;
    persistResult.textContent = `이 브라우저에서는 영속 저장을 쓸 수 없습니다.\n${error.message ?? error}`;
  }

  const source = await fetch('src/main.py').then((response) => response.text());
  // 이 손잡이는 페이지가 사는 동안 계속 쓰므로 일부러 안 놓는다.
  pyGlobals = pyodide.runPython(`${source}\n\nglobals()`);
  call('setup');
  done();
  bootBox.textContent = '준비됐습니다. 파일을 올리고 내려받아 보세요.';

  refreshButton.addEventListener('click', () => showTree());
  uploadInput.addEventListener('change', (event) => upload(pyodide, event));
  downloadButton.addEventListener('click', () => download(pyodide));
  saveSyncButton.addEventListener('click', () => savePersistent(pyodide, true));
  saveNoSyncButton.addEventListener('click', () => savePersistent(pyodide, false));
  reloadButton.addEventListener('click', () => location.reload());
  unpackButton.addEventListener('click', () => unpack(pyodide));

  for (const control of [refreshButton, uploadInput, noteInput, downloadButton, unpackButton]) {
    control.disabled = false;
  }
  if (persistReady) {
    saveSyncButton.disabled = false;
    saveNoSyncButton.disabled = false;
  }

  showTree();
  if (persistReady) showPersistState();
}

/**
 * IDBFS 를 읽거나 쓴다.
 *
 * IndexedDB 가 비동기라서 콜백으로 온다. Promise 로 감싸 두면 부르는 쪽이 편하다.
 * fromDb 가 true 면 저장된 것을 읽어 오고, false 면 지금 것을 저장한다.
 */
function syncfs(pyodide, fromDb) {
  return new Promise((resolve, reject) => {
    pyodide.FS.syncfs(fromDb, (error) => (error ? reject(error) : resolve()));
  });
}

function call(name, ...args) {
  const fn = pyGlobals.get(name);
  try {
    return fn(...args);
  } finally {
    fn.destroy();
  }
}

/** 파이썬이 훑은 목록을 표로 옮긴다. 돌려받은 것도 손잡이라 놓아 준다. */
function showTree() {
  treeBody.replaceChildren();
  for (const root of ROOTS) {
    const rows = call('list_tree', root);
    try {
      addRow(root, rows.length ? '(폴더)' : '(비어 있거나 아직 없음)', true);
      for (const row of rows) {
        // 반복이 끝나면 알아서 회수되는 빌린 손잡이다. 따로 놓아 줄 필요가 없다.
        const [path, size, isDir] = row.toJs();
        addRow(`  ${path.slice(root.length + 1)}`, isDir ? '(폴더)' : formatBytes(size), isDir);
      }
    } finally {
      rows.destroy();
    }
  }
}

function addRow(label, size, muted) {
  const row = document.createElement('tr');
  const head = document.createElement('th');
  head.scope = 'row';
  head.textContent = label;
  if (muted) head.classList.add('dir');
  const cell = document.createElement('td');
  cell.textContent = size;
  row.append(head, cell);
  treeBody.append(row);
}

/** 사용자가 고른 파일을 파일시스템에 넣는다. */
async function upload(pyodide, event) {
  const file = event.target.files?.[0];
  if (!file) return;
  uploadResult.replaceChildren();

  try {
    // File 을 바이트로 바꿔 그대로 쓴다. 이 자리가 브라우저와 파이썬이 만나는 곳이다.
    const bytes = new Uint8Array(await file.arrayBuffer());
    // 이름을 그대로 붙이지 않는다. .. 가 들어 있으면 /work 밖에 쓰게 된다.
    // 파일 선택기는 그런 이름을 주지 않지만, 드래그앤드롭이나 원격에서 온 이름은 다르다.
    const path = `${WORK}/${file.name.split('/').pop()}`;
    pyodide.FS.writeFile(path, bytes);
    uploadResult.textContent = call('describe', path);
    showTree();
  } catch (error) {
    renderPythonError(uploadResult, error);
  } finally {
    // 비워 두지 않으면 같은 파일을 다시 골랐을 때 change 가 오지 않는다.
    event.target.value = '';
  }
}

/** 파이썬이 만든 파일을 내려받는다. */
function download(pyodide) {
  const path = call('make_report', noteInput.value);
  const bytes = pyodide.FS.readFile(path);

  // Blob 으로 감싸 임시 주소를 만들고, 보이지 않는 링크를 눌러 준다.
  const url = URL.createObjectURL(new Blob([bytes], { type: 'text/plain' }));
  const link = document.createElement('a');
  link.href = url;
  link.download = 'report.txt';
  link.click();
  // 주소는 페이지가 사는 동안 남으므로 다 쓰면 놓아 준다.
  URL.revokeObjectURL(url);

  downloadResult.textContent = `${path} 를 만들어 내려받았습니다.\n\n${new TextDecoder().decode(bytes)}`;
  showTree();
}

async function savePersistent(pyodide, withSync) {
  const name = withSync ? 'synced.txt' : 'forgot.txt';
  const path = call(
    'write_persistent',
    name,
    withSync ? '저장까지 했습니다' : '저장은 안 했습니다',
  );

  if (withSync) {
    await syncfs(pyodide, false);
    persistResult.textContent = `${path} 를 쓰고 syncfs 까지 불렀습니다. 새로고침해도 남습니다.`;
  } else {
    persistResult.textContent = `${path} 를 쓰기만 했습니다. syncfs 를 안 불렀으니 새로고침하면 사라집니다.`;
  }
  showTree();
}

/**
 * 페이지를 열었을 때 영속 폴더에 무엇이 남아 있는지 알려 준다.
 *
 * 반복 도중에 값으로 바꿔 담는 것이 중요하다. 파이썬 목록을 훑을 때 나오는 항목은
 * 반복이 끝나면 회수되는 빌린 손잡이라서, 배열에 모아 뒀다가 나중에 쓰면 이미 죽어 있다.
 */
function showPersistState() {
  const rows = call('list_tree', PERSIST);
  const names = [];
  try {
    for (const row of rows) {
      const [path] = row.toJs();
      names.push(path.slice(PERSIST.length + 1));
    }
  } finally {
    rows.destroy();
  }

  persistResult.textContent = names.length
    ? `지난번에 저장해 둔 것: ${names.join(', ')}`
    : '영속 폴더가 비어 있습니다.';
}

/** 파이썬이 만든 zip 을 Pyodide 가 파일시스템에 푼다. */
function unpack(pyodide) {
  const made = call('make_zip');
  try {
    // unpackArchive 는 ArrayBuffer 나 그 뷰를 받는다. 손잡이를 그대로 주면 안 된다.
    // .buffer 를 꺼내면 부분 뷰일 때 엉뚱한 데이터를 넘기게 되므로 뷰째로 준다.
    pyodide.unpackArchive(made.toJs(), 'zip', { extractDir: UNPACKED });
  } finally {
    made.destroy();
  }

  unpackResult.textContent = 'zip 을 만들어 /unpacked 에 풀었습니다. 아래 목록을 보세요.';
  showTree();
}
